internal/server/deviceserver.go

```diff
@@ -15,18 +20,47 @@ type deviceServer struct {
 	pb.UnimplementedWorkerServer
 	configManager       *configuration2.Manager
 	registrationManager *registration2.Registration
+	ansibleManager      *ansible.AnsibleManager
 }
 
-func NewDeviceServer(configManager *configuration2.Manager, registrationManager *registration2.Registration) *deviceServer {
+func NewDeviceServer(configManager *configuration2.Manager, registrationManager *registration2.Registration, ansibleManager *ansible.AnsibleManager) *deviceServer {
 	return &deviceServer{
 		configManager:       configManager,
 		registrationManager: registrationManager,
+		ansibleManager:      ansibleManager,
 	}
 }
 
 // Send implements the "Send" method of the Worker gRPC service.
 func (s *deviceServer) Send(ctx context.Context, d *pb.Data) (*pb.Receipt, error) {
 	go func() {
+		//check if it is an ansible playbook message
+
+		if x, found := d.GetMetadata()["ansible-playbook"]; found && x == "true" {
```

Comment:
How do we make sure not to execute the same playbook over and over again after each call to `GetUpdates` is invoked? 
After all, if the playbook is part of device-configuration, we'll be sending that information few times per-minute which isn't desired to execute over and over (even if idempotent).
Also, which feedback will be sent back on successful or failed playbook execution?

Reply:
> How do we make sure not to execute the same playbook over and over again 
The idea is that it's the operator's responsibility to know whether or not to send the playbook to the device. Makes sense for you?

>which feedback will be sent back on successful or failed playbook execution
The device will send the `ansible events` as results either to the operator or the redhat console ingress service.

Reply:
I think we need to make sure to execute the same playbook twice in parallel.
This needs to be handled on the device side.

From declarative API perspective, the playbook will remain part of the CR.
However, the device-operator api may omit is as part of `get-updates` response if not needed.
And yet, there need to be a definition under which condition not to send a playbook that was already executed.
(reattempt on failure, not to resend a playbook that was completed successfully, assuming it wasn't changed).
This needs to be well defined and documented.

Reply:
> I think we need to make sure to execute the same playbook twice in parallel.

@masayag Any suggestion on how to do this? An in memory map with playbook in execution is acceptable? Or there are other mechanisms for this in the device?

Reply:
What will happen to the executed playbook if the device-worker process is died and restarted?
If it terminates the playbook, we might not have information about its completion.
Are there any on-disk resources created in this flow? Or any way to obtain playbook status outside of the device-worker?
The recommendation will be influenced by the answers to the above, since we'd might want to be able to report playbook execution status if it was cancelled prematurely or even to re-attempt running it.
It can be achieve by keeping track of it on disk, and reflect the status while not completed using in-memory map (since there isn't really need to read it from disk every time if nothing changed).

Reply:
and to complete the other side, while we're running a playbook, we should dismiss the same playbook on the next 'get-updates' call.

Reply:
> Are there any on-disk resources created in this flow? 

No

> Or any way to obtain playbook status outside of the device-worker?

AFAIK, No. 

> and to complete the other side, while we're running a playbook, we should dismiss the same playbook on the next 'get-updates' call.

Is the `get-updates` call this https://github.com/project-flotta/flotta-operator/blob/389e8a025bb6525fb563f80d00c49ccb71f419b3/internal/yggdrasil/yggdrasil.go#L142 ?

Reply:
yes, that's the one.